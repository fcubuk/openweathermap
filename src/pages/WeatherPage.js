import React, { Component } from 'react';
import CSSModules from 'react-css-modules';
import FindForecastForm from '../components/FindForecastForm/FindForecastForm';
import Forecast from '../components/Forecast/Forecast';

import styles from './WeatherPage.scss';
import { getCities, getWeatherData } from "../services/contentService";

const TEMPERATURE_UNITS_CELCIUS = 'metric';
// const TEMPERATURE_UNITS_FAHRENHEIHT = 'imperial';
const ERROR_FETCH_DATA = 'Error ferching forecast data.';

class WeatherPage extends Component {
  state = {
    locations: null,
    location_id: undefined,
    selectedTemperatureUnit: TEMPERATURE_UNITS_CELCIUS,
    forecasts: null,
    isLoading: false,
    error: null
  }

  handleForecastOptions = ({ location_id }) => {
    if (this.state.location_id === location_id) {
      return;
    }
    this.setState({ isLoading: true });
    getWeatherData(location_id, TEMPERATURE_UNITS_CELCIUS)
      .then(forecasts => {
        // console.log('Weather Data:', forecasts);
        this.setState({ location_id, forecasts, isLoading: false, error: null });
      })
      .catch(error =>
        this.setState({ isLoading: false, error: `${ERROR_FETCH_DATA}\n${error}` })
      );
  }

  handleTemperatureUnitChange = (selectedTemperatureUnit) => {
    if (this.state.selectedTemperatureUnit !== selectedTemperatureUnit) {
      this.setState({ selectedTemperatureUnit });
    }
  }

  componentDidMount() {
    if (!this.state.location) {
      getCities().then(locations => {
        if (!!locations) {
          this.setState({ locations })
        }
      });
    }
  }
  

  render() {
    const { isLoading, error, forecasts, locations, location_id, selectedTemperatureUnit } = this.state;
    return (
      <div className={styles['container']}>
        <h1>Weather Page</h1>
        <FindForecastForm
          locations={locations}
          handleForecastOptions={this.handleForecastOptions}
          handleTemperatureUnitChange={this.handleTemperatureUnitChange}
          locationID={location_id}
          selectedTemperatureUnit={selectedTemperatureUnit}
        />
        {isLoading &&
          <div>Loading...</div>
        }
        {!!error ?
          (<ul className={styles['error']}>
            <li>{error}</li>
          </ul>
          ) : (
            <div className={styles['forecasts-container']}>
              {
                (!!forecasts) &&
                <Forecast key={forecasts.location.id} data={forecasts} unit={selectedTemperatureUnit} />
              }
            </div>
          )
        }
      </div>
    )
  }
}

export default CSSModules(WeatherPage, styles);