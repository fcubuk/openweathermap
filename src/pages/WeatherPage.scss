.container {
  padding: 10px 30px 30px 30px;
  .forecasts-container {
    margin-top: 20px;
  }
  .error {
    color: red;
  }
}