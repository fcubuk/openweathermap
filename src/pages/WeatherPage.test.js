/* 
* @jest-environment jsdom
*/
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { spy } from 'sinon';
import WeatherPage from './WeatherPage';

jest.mock('../services/contentService');

describe('<WeatherPage /> component', () => {
  it('should render without data correctly', () => {
    const div = document.createElement('div');
    ReactDOM.render(<WeatherPage />, div);
    ReactDOM.unmountComponentAtNode(div);
    const wrapper = shallow(<WeatherPage />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls componentDidMount() lifecycle method and gets locations data', done => {
    const wrapper = shallow(<WeatherPage />);

    setTimeout(() => {
      const state = wrapper.instance().state;
      const locationsLength = state.locations.length;
      expect(locationsLength).toBe(3)
      done();
    });
  });

  it('calls handleForecastOptions to fetch data', done => {
    const wrapper = shallow(<WeatherPage />);
    // console.log('forecasts', wrapper.state('forecasts'))
    const location_id = 2648110;
    const instance = wrapper.instance();
    setTimeout(() => {
      instance.handleForecastOptions({ location_id });
      // console.log('forecasts ===', wrapper.state('forecasts'))
      expect(wrapper.state('forecasts')).toBeNull()
      done();
    })
  });


});