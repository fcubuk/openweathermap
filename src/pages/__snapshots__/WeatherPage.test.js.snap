// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherPage /> component should render without data correctly 1`] = `
<div
  className="container"
>
  <h1
    key=".0"
  >
    Weather Page
  </h1>
  <FindForecastForm
    handleForecastOptions={[Function]}
    handleTemperatureUnitChange={[Function]}
    key=".1"
    locations={null}
    selectedTemperatureUnit="metric"
  />
  <div
    className="forecasts-container"
    key=".3"
  />
</div>
`;
